4.1: CASSANDRA-19703: Ensure prepared_statement INSERT timestamp precedes eviction DELETE

tolbertam opened the pull request:
Updates SystemKeyspace.writePreparedStatement to accept a timestamp
associated with the Prepared creation time. Using this timestamp
will ensure that an INSERT into system.prepared_statements will
always precede the timestamp for the same Prepared in
SystemKeyspace.removePreparedStatement.

This is needed because Caffeine 2.9.2 may evict an entry as soon
as it is inserted if the maximum weight of the cache is exceeded
causing the DELETE to be executed before the INSERT.

Additionally, any clusters currently experiencing a leaky
system.prepared_statements table from this bug may struggle to
bounce into a version with this fix as
SystemKeyspace.loadPreparedPreparedStatements currently does
not paginate the query to system.prepared_statements, causing heap
OOMs.  To fix this this patch adds pagination at 5000 rows and
aborts loading once the cache size is loaded. This should allow
nodes to come up and delete older prepared statements that may no
longer be used as the cache fills up (which should happen immediately).

This patch does not address the issue of Caffeine immediately evicting
a prepared statement, however it will prevent the
system.prepared_statements table from growing unbounded.  For most users
this should be adequate, as the cache should only be filled when there
are erroneously many unique prepared statements. In such a case we can
expect that clients will constantly prepare statements regardless
of whether or not the cache is evicting statements.

patch by Andy Tolbert; reviewed by Berenguer Blasi and Caleb Rackliffe for CASSANDRA-19703

bereng commented on src/java/org/apache/cassandra/cql3/QueryProcessor.java:
```diff
@@ -155,6 +155,12 @@ private enum InternalStateInstance
 
     public void preloadPreparedStatements()
+    {
+        preloadPreparedStatements(5000);
```
That 5K should be a constant defined somewhere imo

tolbertam replied:
:+1: agreed, especially since its used in multiple places.  I'll introduce a constant.

One thing we've learned since deploying this is that paging does prevent OOMing Cassandra on start up, but the fact that in some cases we preload millions of leaked statements, it can defer startup almost indefinitely.  To get around that we've truncated the prepared_statements table, but have only had to do that in like 2 or 3 out of hundreds of clusters.   I've considered capping the number of statements read and then perhaps truncating the table after, but that feels somewhat wrong to do, especially since the leaking statement issue should go away after this fix, and given this bug has been present for a long time (but exacerbated by the caffeine library change in 4.1+), it didn't feel right to do that.   I think I'll add a warning after ~100k rows are read and leave it at that.

bereng replied:
Right. A warning with an explanation of why that could be a problem would be great. And indeed the bug is present no matter what. It's just that some lib versions make it more visible or less.

tolbertam replied:
I added something for this that logs when we exceed `QueryProcessor.PRELOAD_PREPARED_STATEMENTS_MAX` (1 million).

I thought a little bit more about how I would have wanted this to work, and rather than having it just log and continue plugging away for hours or days, that we just short circuit and stop reading after an excessive amount of rows.

> WARN  [main] 2025-04-12 20:17:30,536 SystemKeyspace.java:1893 - Preloaded 250 prepared statements. This could be an indication that prepared statements leaked prior to CASSANDRA-19703 being fixed. Returning early to prevent indefinite startup.  Consider truncating system.prepared_statements to clear out leaked prepared statements.

The up side here is no user intervention is needed here unless one *really* wants the optimization of prepared statements being preloaded on startup.  If one gets into this situation where you have leaked over 1 million prepared statements, it's likely the operator wasn't aware that they hadn't benefitted from this functionality for quite some time and that misconfigurations in application code had caused a lot of unique prepared statements.

The downside is that the prepared statement table may eventually never be drained without user intervention.  But I consider this a better tradeoff than nodes not starting after hours or days.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +148,103 @@ public void testPstmtInvalidation() throws Throwable
+
+
+    @Test
+    public void testAsyncPstmtInvalidation() throws Throwable
```
Does this deterministically fail without the fix?  Have you explored using byteman to try to register if the right code paths are hit instead i.e.?

tolbertam replied:
Not deterministically because the behavior relies on ordering between competing threads and the underlying caffeine cache behavior, but it reproduces very consistently with the previous implementation.   The larger `statementsToPrepare` is, the more possibility the test will fail.

I will think whether I can test for something more specific than the assertion at line 205 which checks to ensure no statements leaked.  Leaking prepared statements should effectively not be possible with the introduction of the use of client timestamps.

bereng replied:
Rather than doing a sort of E2E test, which will be heavy on CI and execution times, we could target the actual fix itself. I picture this as adding to `testPstmtInvalidation` some byteman checking timestamps were used both in inserts and deletions. But that's just me, just a suggestion.

That way you test the specifics and avoid adding unnecessary CI load.

tolbertam replied:
Thanks for the byteman instrumentation suggestion.  I added something for that which enables the test to ensure that all load timestamps precede delete timestamps and it seems to work quite well.

While it looks like the test is doing quite a bit, it's not actually doing much more than `testPstmtInvalidation` other than preparing statements asynchronously with a concurrency of 10 in an executor and with some additional validation.  The test takes on the order of ~1200ms where testPstmtInvalidation takes ~500ms.

I think there is still some value in it as it is more representative of how prepared statements can be prepared in practice (multiple threads preparing statements concurrently), but I'm also ok with just merging the validation i added here into `testPstmtInvalidation`, let me know what you think :+1:

bereng replied:
OK sgtm. I'd reuse other test cases but we're getting in personal taste territory and I prefer to favor authorship in this cases.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +148,103 @@ public void testPstmtInvalidation() throws Throwable
+    }
+
+    @Test
+    public void testPreloadPreparedStatements() throws Throwable
```
This is just testing statement load but not that paging is being exercised when doing so. Wouldn't we want to check specifically paging is being hit instead with byteman i.e.? Maybe you could reuse any of the other tests to check paging was indeed hit while loading pstmnts

tolbertam replied:
When I wrote this test I was more of the mindset of ensuring that introducing paging didn't put us in position where we missed preloading statements, less about the paging itself, but I think that should definitely be verifiable, will add that :+1:

bereng replied:
You could just add some byteman to `testCachedPreparedStatements` on reload and test the paging is hit there. My 2cts.

tolbertam replied:
I couldn't find a good example of paging tests using byteman to assert that paging actually happened, but I did find a lot of good inspiration in tests using byteman generally to make this work, hopefully the way I went about this looks alright. 

I figured I would keep this test since `testCachedPreparedStatements` only creates a very small amount of prepared statements (~7) and working with on the order of hundreds of statements made it a little easier to validate paging and also validate the additional logic I had to prevent loading past a maximum amount of statements.

tolbertam commented on src/java/org/apache/cassandra/cql3/QueryProcessor.java:
```diff
@@ -101,6 +102,12 @@ public class QueryProcessor implements QueryHandler
+    public static final int PRELOAD_PREPARED_STATEMENTS_FETCH_SIZE = 5000;
+
+    // Maximum number of prepared statements to preload from prepared statement cache.
+    public static final int PRELOAD_PREPARED_STATEMENTS_MAX = 1_000_000;
```
does 1 million feel about right here?  Would there be any value in making this configurable? I can't imagine anyone would actually want to do this, but we could make it a system property if that has some value.

bereng replied:
Thinking about it we don't need to come up with this new limit. We already have a limit on the cache size in MB. Once you hit that it makes no sense to just be trashing the cache. So just drop this and monitor cache size instead. Once you've reached the max cache size you can stop loading. Does it make sense?

tolbertam replied:
:+1: (further discussion [here](https://github.com/apache/cassandra/pull/3917#discussion_r2042139638))

tolbertam commented on src/java/org/apache/cassandra/cql3/QueryProcessor.java:
```diff
@@ -109,15 +116,8 @@ public class QueryProcessor implements QueryHandler
-                                     SystemKeyspace.removePreparedStatement(md5Digest);
-                                 }
-                             }).build();
+                             .removalListener((key, prepared, cause) -> evictPreparedStatement(key, cause))
```
This may look a bit unnecessary, but needed this to instrument byteman to detect when this happens as `removePreparedStatement` is called in other code paths.  I think the code is marginally more concise/unharmfull in any case.

tolbertam commented on doc/modules/cassandra/pages/cql/cql_singlefile.adoc:
```diff
@@ -239,6 +239,20 @@ provide values for `LIMIT`, `TIMESTAMP`, and `TTL` clauses. If anonymous
 bind markers are used, the names for the query parameters will be
 `[limit]`, `[timestamp]`, and `[ttl]`, respectively.
 
+Prepared Statements are cached by cassandra in-memory which can be
```
@bschoening , following up on [your comment](https://issues.apache.org/jira/browse/CASSANDRA-19703?focusedCommentId=17932752&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17932752), how does this look?

In particular I think the second paragraph would be useful in that it emphasizes that one should qualify all values with bind markers to avoid preparing every query which decreases performance for multiple reasons.

bschoening replied:
@tolbertam that's helpful, but until I looked carefully at the code it wasn't clear what the eviction policy was (LRU?) or if bad statements (i.e, not correctly qualified) will eventually be evicted or not. Mentioning Caffeine and the eviction policy used.

How about something like this:  
"To ensure optimal performance, it's important to use bind markers '?' for all non-constant values in your SQL statements. If you include literal values directly in the query instead, each variation will be treated as a unique statement that must be prepared and cached separately. This will soon overflow the prepared statement cache, which is small by design.

When the Caffeine-managed cache reaches its maximum size, older or less frequently used statements are evicted, leading to additional overhead as previously prepared statements must be re-prepared."

tolbertam replied:
That sounds great to me.  I just pushed a change which takes your suggestions as is with a few tweaks, how does that look?

bschoening replied:
@tolbertam looks great, thanks!

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -35,20 +43,32 @@
- 
-    @Test
+
+
```
Why remove this?

tolbertam replied:
this most have been incidentally removed when I was experimenting with byteman annotations, good catch, will add that back.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +162,179 @@ public void testPstmtInvalidation() throws Throwable
             {
+                assertEquals("Number of statements in table and in cache don't match", numberOfStatementsInMemory(), numberOfStatementsOnDisk());
+
+                // prepare a more statements to trigger more evictions
```
`a more` is a typo?

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +162,179 @@ public void testPstmtInvalidation() throws Throwable
+    @Test
+    @BMRules(rules= {
+    @BMRule(name = "CaptureWriteTimestamps",
+    targetClass = "SystemKeyspace",
```
Formatting

tolbertam replied:
agree that it looks awkward (for some reason me IDE really wanted to coerce me into it).  Pushed something  which looks a bit better hopefully.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +162,179 @@ public void testPstmtInvalidation() throws Throwable
+    // page size passed to preloadPreparedStatements
+    private static final int PRELOAD_PAGE_SIZE = 100;
+    // recorded page invocations in preloadPreparedStatements
+    private static int PAGE_INVOCATIONS = 0;
```
This should probably be volatile and lowercase as it is not a constant. Both should be defined at the top where we define such stuff in all tests, just for consistency.

tolbertam replied:
sounds good, updated and also changed to AtomicInteger since we're incrementing an existing value.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +162,179 @@ public void testPstmtInvalidation() throws Throwable
+    @SuppressWarnings("unused")
+    private static void nextPageReadQuery(int pageSize)
+    {
+        if (pageSize == PRELOAD_PAGE_SIZE)
```
This is super hacky and brittle. Can't you access through `$0` the query member of `PartitionRangeQueryPager` to check it is indeed the right CF and that should be enough?

tolbertam replied:
improved this by passing in the query and validating the table name.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +162,179 @@ public void testPstmtInvalidation() throws Throwable
+
+    @Test
+    @BMRule(name = "CapturePageInvocations",
+    targetClass = "PartitionRangeQueryPager",
```
Formatting

bereng commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1863,16 +1865,37 @@ public static void resetPreparedStatements()
                                 row.has("logged_keyspace") ? row.getString("logged_keyspace") : null))
                 counter++;
+
+            if (counter >= max)
```
Here you could check cache size every page, not every record for perf reasons, dropping `max` which we wouldn't need anymore and would not adapt to the cache size which is user configurable..

tolbertam replied:
that would make a ton more sense, thanks for the suggestion :+1:

tolbertam replied:
Updated this to short circuit on detection of cache eviction, I still need to update this to do the evaluation on page boundaries which I haven't figured out how to detect yet (need to find something like `isExhausted()` on page boundary which I haven't stumbled into yet).

tolbertam replied:
Think best I could do is just check every X rows and then check to see if we've started evicting, pushed something for that and hopefully that looks good!

bereng replied:
Added a note of checking the cache size directly rather than the evictions counter which is not deterministic.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +172,172 @@ public void testPstmtInvalidation() throws Throwable
+                     targetLocation = "AT INVOKE executeInternal",
+                     action = "org.apache.cassandra.cql3.PstmtPersistenceTest.preparedStatementLoadTimestamps.put($key, $timestamp);"
+             ),
+            @BMRule(name = "CaptureEvictTimestamps",
```
Nit: Not aligned by one space

tolbertam replied:
🤦 good catch, will fix this shortly.

bereng commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1863,16 +1864,38 @@ public static void resetPreparedStatements()
                 counter++;
+
+            if (counter % pageSize == 0
+                && QueryProcessor.metrics.preparedStatementsEvicted.getCount() - initialStatementsEvicted > 0)
```
We can't rely on the evictions counter imo. We have code in QueryProcessor invalidating entries, other proceses could be invalidating stuff, also iirc on collisions evictions where triggered inside caffeine, etc. Why not just check the size?
 ```
eviction = cache.policy().eviction().orElseThrow()
long currentWeight = eviction.weightedSize().orElseThrow()
```
Give it an extra 5% (?) to accommodate for Caffeine async evictions that may occasionally go over the limit and we should be ok

tolbertam replied:
I hadn't considered that as this method would be called on startup before listening on CQL, but there could always be internal preparations and statements evicted for other reasons.  I agree that testing against the cache size would probably be more reliable.  Will fix that :+1:

tolbertam replied:
Was just thinking this through and I was cautiously worried about two things:

1. Incidentally warning the user that they might have hit a leak, when they really haven't (e.g. if cache was full on restart and we detected it was full and we 5% wasn't enough of a buffer)
2. Not detecting the cache is full because 5% is too big of a buffer and paging indefinitely.

But I just realized that one thing we can use in our favor is that we only evaluate this on paging boundaries, so its exceedingly unlikely that either of those 2 cases would be a problem with a page size of 5000.  Therefore, I think it's ok to do this, and that we also don't need a 5% buffer, as evaluating on page boundaries acts as a buffer in itself.  Will push something shortly that will should work I hope.

tolbertam replied:
Pushed something that I think this should be sufficient, but do have a small inkling of doubt.

We could consider adding a 5% or so buffer as you suggested, but by under-compensating (e.g. `currentSize >= maximumSize * 0.95`) so we can handle the case of the cache being evicted right below it's capacity.

I suspect if we page through leaked statements we will eventually hit a case where the cache is briefly above its capacity and it will cut off.

If we do add an under-compensation, it's possible that we will erroneously stop and log a warning which may not be correct.  On the other hand we could always adjust that warning to indicate that it's possible we stopped because the cache was close to full.

In the usual case, the prepared statement cache should not really approach close to getting full unless non-fully qualified statements are being prepared anyways, so it might be useful to warn in that case as well ("Detected that the prepared statement cache was close to full on preload, please investigate whether your statements being prepared are fully qualified").  Thoughts?

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +172,172 @@ public void testPstmtInvalidation() throws Throwable
+     * Invoked by byteman whenever QueryProcessor.evictPreparedStatement is called in testAsyncPstmtInvalidation
+     */
+    @SuppressWarnings("unused")
+    private static void recordEvictionTimestamp(MD5Digest key)
```
Could this be removed and call `put` directly as you do with `preparedStatementLoadTimestamps`?

tolbertam replied:
🤔 maybe, i'll see if I can make it work.  The main reason I did this was because we need to use `ClientState.getTimestamp()` to generate a timestamp to record when this was invoked (since we don't use explicit timestamps on the delete)  But perhaps I can make this works with lambdas, or alternative change `evictPreparedStatement` to require a client timestamp.

tolbertam replied:
was able to make this work, thanks!

bereng commented:
IIUC here we're fixing 3 things:
- Ordering of timestamps preventing the leak
- Paged reading when preloading to further soften any leaks
- Stop preloading when cache is already full to prevent trashing the cache upon leaks

tolbertam commented on doc/modules/cassandra/pages/cql/cql_singlefile.adoc:
```diff
@@ -239,6 +239,28 @@ provide values for `LIMIT`, `TIMESTAMP`, and `TTL` clauses. If anonymous
+The cache is also persisted to the `system.prepared_statements` table
+so it can be preloaded into memory on startup.
+
+To ensure optimal performance, it's important to use a bind `<variable>`
```
Note: opted to use `<variable>` over bind marker to match the preceding documentation.

bereng commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1863,16 +1864,37 @@ public static void resetPreparedStatements()
+
+            if (counter % pageSize == 0 && QueryProcessor.isPreparedStatementCacheFull())
+            {
+                // In the event that we detect that the prepared statement cache has filled up, return early to prevent
```
I like this explanation a lot +1

bereng replied:
I am only worried here we'd detect only the problem if _both_ we're on a page boundary and caffeine happens to be reporting full. Which could be a racy or rare event. But if you say it works for you in production so be it. It's better than nothing and we can always tweak if we reports from the field suggesting a change.

tolbertam replied:
I think it doesn't have to be too precise.  We may in fact page several times more than what would be ideal, but the important thing is that we stop and don't page indefinitely.

We could make it more sensitive by changing the calculation to be some value less than the maximum weight (something like 95%).  The downside of doing that is we may drop prepared statements that we may have been able to cache otherwise and would also notify the user that the cache may be leaky when it in fact that isn't the issue.

tolbertam replied:
> Aren't we missing a test we stop loading pstmnts once cache is full? Maybe you could use byteman to capture the call to the log warn method. That with a modified yaml to have a smaller pstmnts cache somehow should do the trick. I think it's one important feature we're not testing unless I am missing sthg.

I'm glad you brought this up as it pushed me to exercise how effective this logic could be, and apparently it was not very effective as is!

We deployed an early version of this fix in late 2024/early 2025, at which time we didn't have any short circuiting logic, so we would just page indefinitely (which was better than OOMing but still not great).  When we encountered this, we would simply just truncate the `system.prepared_statements` table.  So this particular piece of logic hadn't really been put through its paces.

After writing a test (`testPreloadPreparedStatementsUntilCacheFull`) I found that it was *very* difficult to encounter a full cache as Caffeine, while it allows going past the maximum weight, is very effective at evicting quickly below it.  Therefore, I think we need to add some kind of negative buffer.   I found that testing against 99% of the cache size was very effective.

To avoid potentially logging that there might be a leak when we previously just had a nearly full cache, we now allow paging one addition time (`possibleLeakDetected`), at which point if we still haven't exhausted the result set, we'll log the message and exit earlier.

In terms of validating where we stop once the cache is full, I found that we could easily detect this by updating `preloadPreparedStatements` to return an `int` and then comparing the total number of rows returned against the maximum amount we should expect (I added some github comments to the test I added to expand on this more).  I think this is more precise than testing against the log entry, which also does indeed get logged in the test:

> WARN  [main] 2025-05-06 22:36:42,581 SystemKeyspace.java:1888 - Detected prepared statement cache filling up during preload after preparing 492 statements. This could be an indication that prepared statements leaked prior to CASSANDRA-19703 being fixed. Returning early to prevent indefinite startup. Consider truncating system.prepared_statements to clear out leaked prepared statements.

Hopefully that all sounds right!

bereng replied:
> I found that it was very difficult to encounter a full cache

That is exactly what I was referring to. We were trying to detect a full cache _and_ only at page boundaries. Which sounded very improbable at face value knowing caffeine.

bereng replied:
That 1% you added is like the 5% I was suggesting iiuc.

But you know sometimes I'm just an idiot. As every pstmnt comes with it's size precomputed now, you can add those as you load them. Then compare against the max weight of the cache trashing and simplifying all this. You stop loading when the accumulated size > caffeine max weight.

Trash all those page boundaries, cache full methods, counters, % guards,...

Apologies about the noise. I don't know how I didn't think about it before :shrug:

tolbertam replied:
That does sound like a more effective way of doing this, thanks!  I just pushed 42496b5, hopefully that looks good.  
I kept the logic there that only tests this on page boundaries because I found in testing that Caffeine's caching behavior allows us to load past the cache size, so in order to prevent erroneously logging on startup its worth having a buffer before we log and stop iterating.

bereng replied:
You can trash all that as you have a config property for the max cache size. Commented inline.

bereng commented:
Aren't we missing a test we stop loading pstmnts once cache is full? Maybe you could use byteman to capture the call to the log warn method. That with a modified yaml to have a smaller pstmnts cache somehow should do the trick. I think it's one important feature we're not testing unless I am missing sthg.

tolbertam commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +172,200 @@ public void testPstmtInvalidation() throws Throwable
+        // Ensure the page size we use is less than the number of statments it took to load the cache.
+        int pageSize = statementsLoadedBeforeFull / 10;
+        // We should only expect to load how many statements we were able to load before filling the cache
+        // + a buffer of a page size + another possible page.
```
we allow a buffer of 2 pages for:

one page to hit the boundary where we detect we might have leaked + 
additional page that we are allowed to read before exiting early.

bereng replied:
I would trash that and just loop and break when you go over the limit. it's just 2 local variable reads on startup and you drop unnecessary complexity of the code.

tolbertam commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1862,17 +1865,61 @@ public static void resetPreparedStatements()
     }
 
-    public static int loadPreparedStatements(TriFunction<MD5Digest, String, String, Boolean> onLoaded)
+    public static int loadPreparedStatements(TriFunction<MD5Digest, String, String, Prepared> onLoaded)
```
Hopefully it's ok to change the type signature here?  I don't expect this method is used outside of the codebase itself and tests.

bereng replied:
We need it to pull the size so it is justifiable imo

tolbertam commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1862,17 +1865,61 @@ public static void resetPreparedStatements()
+            {
+                if (possibleLeakDetected)
+                {
+                    logger.warn("Detected prepared statement cache filling up during preload after preparing {} " +
```
What this looks like:

> WARN  [main] 2025-05-07 09:30:16,724 SystemKeyspace.java:1898 - Detected prepared statement cache filling up during preload after preparing 492 statements (1.17 MiB of 1 MiB cache size). This could be an indication that prepared statements leaked prior to CASSANDRA-19703 being fixed. Returning early to prevent indefinite startup. Consider truncating system.prepared_statements to clear out leaked prepared statements.

bereng replied:
Nice :-)

tolbertam replied:
new variant:

> WARN  [main] 2025-05-08 21:31:54,578 SystemKeyspace.java:1904 - Detected prepared statement cache filling up during preload after preparing 463 statements (loaded 1.1 MiB with prepared_statements_cache_size being 1 MiB). This could be an indication that prepared statements leaked prior to CASSANDRA-19703 being fixed. Returning early to prevent indefinite startup. Consider truncating system.prepared_statements to clear out leaked prepared statements.

bereng replied:
I think it now reads much cleaner :-)

bereng commented on src/java/org/apache/cassandra/cql3/QueryProcessor.java:
```diff
@@ -131,6 +129,28 @@ public class QueryProcessor implements QueryHandler
+     * @return The maximum capacity, in bytes, of the prepared statements cache.  If no capacity is established,
+     * {@link Long#MAX_VALUE} is returned.
+     */
+    public static long getPreparedStatementsCacheMaximumCapacity()
```
Trash this. You already have `DatabaseDescriptor.getPreparedStatementsCacheSizeMiB()`

tolbertam replied:
removed

bereng commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1862,17 +1865,61 @@ public static void resetPreparedStatements()
+        long preparedBytesLoaded = 0L;
+        long cacheCapacity = QueryProcessor.getPreparedStatementsCacheMaximumCapacity();
         for (UntypedResultSet.Row row : resultSet)
         {
```
This is extremely hard to read imo. I understand your worried about pulling the size from the cache etc. But that is not the case anymore as we have that as a config value. Just loop and break KISS

tolbertam replied:
:+1: hopefully it's more concise now that I've removed the validation at paging boundaries.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +172,203 @@ public void testPstmtInvalidation() throws Throwable
+                cnt++;
+            }
+
+            if (statementsLoadedBeforeFull == -1) {
```
braces formatting

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +172,203 @@ public void testPstmtInvalidation() throws Throwable
+            // load prepared statements until we detect any eviction, on eviction we assume cache was filled.
+            while (QueryProcessor.metrics.preparedStatementsEvicted.getCount() == initialEvicted)
+            {
+                prepareStatement("INSERT INTO %s (key, val) VALUES (?, ?) USING TIMESTAMP " + cnt, clientState);
```
No variables bound?

tolbertam replied:
Unless i'm misunderstanding, it shouldn't be necessary as we just need the prepared statement query.

bereng commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +172,203 @@ public void testPstmtInvalidation() throws Throwable
+    }
+
+    @Test
+    public void testPreloadPreparedStatementsUntilCacheFull()
```
An alternative no relying on cache evictions could be:
```
While(insertedPstmntsSize < Config.pStmntCacheSize)
{
    insert pstmnt
    insertedPstmntsSize += pstmnt.size
    insertedPstmntsCount ++
}

count = preloadPreparedStaments()
Assert counts match
Assert warn was issued
```

tolbertam replied:
that sounds better, thanks for the suggestion!  I've updated to match this short of the logging validation.

I'm not finding a lot of precedence around validating logging other than for logging specific functionality (`AuditLoggerAuthTest`, `StatusLoggerTest`, `NoSpamLoggerTest`).  Since we are able to concisely validate that we short circuit preloading statements it feels adequate to me.

tolbertam commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -178,7 +400,6 @@ private MD5Digest prepareStatement(String stmt, ClientState clientState)
 
     private MD5Digest prepareStatement(String stmt, String keyspace, String table, ClientState clientState)
     {
-        System.out.println(stmt + String.format(stmt, keyspace + "." + table));
```
Not sure how that println got in there but it's been there for 4 years apparently, figure it's ok to remove.

bereng replied:
That's amazing lol. Good catch

tolbertam commented on src/java/org/apache/cassandra/db/SystemKeyspace.java:
```diff
@@ -1862,17 +1866,52 @@ public static void resetPreparedStatements()
+        // to avoid logging early.
+        long cacheSize = new DataStorageSpec.LongBytesBound(DatabaseDescriptor.getPreparedStatementsCacheSizeMiB(),
+                                                            DataStorageSpec.DataStorageUnit.MEBIBYTES).toBytes();
+        long preparedBytesLoadThreshold = (long) (cacheSize * 1.1);
```
I found that we still need a small buffer to avoid erroneously logging.  In my testing found that we often loaded a statement or two past the cache size. A 10% buffer should be more than adequate.

bereng replied:
We better align to the code when the cache's size is initialized at the top of the class:
`cacheSize=capacityToBytes(DatabaseDescriptor.getPreparedStatementsCacheSizeMiB())`

bereng commented on src/java/org/apache/cassandra/cql3/QueryProcessor.java:
```diff
@@ -131,6 +128,16 @@ public class QueryProcessor implements QueryHandler
                     DatabaseDescriptor.getPreparedStatementsCacheSizeMiB());
     }
 
+    public static void evictPreparedStatement(MD5Digest key, RemovalCause cause)
```
I would add annotation VisibleForTesting here? It is not intended to be called iiuc

tolbertam replied:
:+1:, looking at it again now I just realized this doesn't even need to be public, so I'll reduce the visibility to private. 

We ended up only needing this as a separate method to ensure it gets invoked in `PstmtPersistenceTest` and apparently byteman is still able to instrument around it with private visibility.

bereng commented:
LGTM modulus the 2 minor issues I mentioned last. I would still heavily compact junits but that's just me. We need another C* committer to +1 this now right?

tolbertam commented:
Pushed some changes which hopefully addresses the remaining feedback, hopefully everything looks ok!  I merged in recent changes from 4.1 and ran a [full CI run](https://issues.apache.org/jira/browse/CASSANDRA-19703?focusedCommentId=17950821&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17950821).  I'll find a second reviewer tomorrow.

bereng commented on src/java/org/apache/cassandra/cql3/QueryProcessor.java:
```diff
@@ -128,7 +131,8 @@ public class QueryProcessor implements QueryHandler
     }
 
-    public static void evictPreparedStatement(MD5Digest key, RemovalCause cause)
+
```
Extra blank line

tolbertam replied:
:+1: sorry about that, fixed.

tolbertam committed:
Ensure prepared_statement INSERT timestamp precedes eviction DELETE

Updates SystemKeyspace.writePreparedStatement to accept a timestamp
associated with the Prepared creation time. Using this timestamp
will ensure that an INSERT into system.prepared_statements will
always precede the timestamp for the same Prepared in
SystemKeyspace.removePreparedStatement.

This is needed because Caffeine 2.9.2 may evict an entry as soon
as it is inserted if the maximum weight of the cache is exceeded
causing the DELETE to be executed before the INSERT.

Additionally, any clusters currently experiencing a leaky
system.prepared_statements table from this bug may struggle to
bounce into a version with this fix as
SystemKeyspace.loadPreparedPreparedStatements currently does
not paginate the query to system.prepared_statements, causing heap
OOMs.  To fix this this patch adds pagination at 5000 rows and
aborts loading once the cache size is loaded. This should allow
nodes to come up and delete older prepared statements that may no
longer be used as the cache fills up (which should happen immediately).

This patch does not address the issue of Caffeine immediately evicting
a prepared statement, however it will prevent the
system.prepared_statements table from growing unbounded.  For most users
this should be adequate, as the cache should only be filled when there
are erroneously many unique prepared statements. In such a case we can
expect that clients will constantly prepare statements regardless
of whether or not the cache is evicting statements.

patch by Andy Tolbert; reviewed by Berenguer Blasi and Caleb Rackliffe for CASSANDRA-19703

tolbertam commented:
went ahead and rebased and squashed the commits as I had updated from trunk previously which messed with the history a bit.  Made no changes and ensured a clean diff before pushing.

maedhroz commented on CHANGES.txt:
```diff
@@ -5,6 +5,7 @@
  * Optionally skip exception logging on invalid legacy protocol magic exception (CASSANDRA-19483)
  * Fix SimpleClient ability to release acquired capacity (CASSANDRA-20202)
  * Fix WaitQueue.Signal.awaitUninterruptibly may block forever if invoking thread is interrupted (CASSANDRA-20084)
+ * Ensure prepared_statement INSERT timestamp precedes eviction DELETE (CASSANDRA-19703)
```
nit: I think new entries go at the top

tolbertam commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -21,33 +21,65 @@
         QueryProcessor.clearPreparedStatements(false);
     }
- 
+
```
Looked odd, but previous line had a space in it; removed second new line.

tolbertam commented on test/unit/org/apache/cassandra/cql3/PstmtPersistenceTest.java:
```diff
@@ -142,19 +174,209 @@ public void testPstmtInvalidation() throws Throwable
+        // should detect a possible leak and defer paging indefinitely by returning early in preloadPreparedStatements.
+        int statementsLoadedWhenFull = -1;
+        long accumulatedSize = 0;
+        long cacheSize = new DataStorageSpec.LongBytesBound(DatabaseDescriptor.getPreparedStatementsCacheSizeMiB(),
```
Could use QueryProcessor.PREPARED_STATEMENT_CACHE_SIZE_BYTES instead now that we have a constant for this.  will fix.

bereng commented:
LGTM, we need PRs and CI for the other branches now.

tolbertam commented:
thanks :+1: will go ahead and squash into a single commit, and kick off CI against 4.1/5.0/trunk branches.

tolbertam commented:
squashed with no changes.

bereng commented:
I think we need 4.0 as well

tolbertam commented:
> I think we need 4.0 as well

Technically the bug is not likely to be encountered on 4.0 as Caffeine is running an older version that behaves differently, but it wouldn't hurt to do so, will create a branch.

tolbertam commented:
Added two small changes (c85d156 and 3c80c2e) which I have squashed on the other branches.  I have added test results to [the jira](https://issues.apache.org/jira/browse/CASSANDRA-19703?focusedCommentId=17952487&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17952487).  Can squash into a single commit assuming everything looks good.

tolbertam commented:
squashed with no changes.
